debian/grub/20_acrn-board-inspector

```diff
@@ -0,0 +1,269 @@
+  LINUX_ROOT_DEVICE=PARTUUID=${GRUB_DEVICE_PARTUUID}
+else
+  LINUX_ROOT_DEVICE=UUID=${GRUB_DEVICE_UUID}
+fi
+
+case x"$GRUB_FS" in
+    xbtrfs)
+	rootsubvol="`make_system_path_relative_to_its_root /`"
+	rootsubvol="${rootsubvol#/}"
+	if [ "x${rootsubvol}" != x ]; then
+	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
+	fi;;
+    xzfs)
+	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || true`
+	bootfs="`make_system_path_relative_to_its_root / | sed -e "s,@$,,"`"
+	LINUX_ROOT_DEVICE="ZFS=${rpool}${bootfs%/}"
+	;;
+esac
+
+if [ -x /lib/recovery-mode/recovery-menu ]; then
+    GRUB_CMDLINE_LINUX_RECOVERY=recovery
+else
+    GRUB_CMDLINE_LINUX_RECOVERY=single
+fi
+
+linux_entry ()
+{
+  os="$1"
+  version="$2"
+  args="$3"
+
+  if [ -z "$boot_device_id" ]; then
+      boot_device_id="$(grub_get_device_id "${GRUB_DEVICE}")"
+  fi
+  echo "menuentry '$(echo "$os" | grub_quote)' ${CLASS} \$menuentry_id_option 'gnulinux-acrn-board-inspector-$boot_device_id' {"
+
+  # Use ELILO's generic "efifb" when it's known to be available.
+  # FIXME: We need an interface to select vesafb in case efifb can't be used.
+  if [ "x$GRUB_GFXPAYLOAD_LINUX" = x ]; then
+      echo "	load_video"
+  else
+      if [ "x$GRUB_GFXPAYLOAD_LINUX" != xtext ]; then
+	  echo "	load_video"
+      fi
+  fi
+  if [ "x$GRUB_GFXPAYLOAD_LINUX" != x ]; then
+      echo "	gfxmode \$linux_gfx_mode"
+  fi
+
+  echo "	insmod gzio"
+
+  if [ x$dirname = x/ ]; then
+    if [ -z "${prepare_root_cache}" ]; then
+      prepare_root_cache="$(prepare_grub_to_access_device ${GRUB_DEVICE} | grub_add_tab)"
+    fi
+    printf '%s\n' "${prepare_root_cache}"
+  else
+    if [ -z "${prepare_boot_cache}" ]; then
+      prepare_boot_cache="$(prepare_grub_to_access_device ${GRUB_DEVICE_BOOT} | grub_add_tab)"
+    fi
+    printf '%s\n' "${prepare_boot_cache}"
+  fi
+  message="$(gettext_printf "Loading Linux %s ..." ${version})"
+  cat << EOF
+	echo	'$(echo "$message" | grub_quote)'
+EOF
+  if test -d /sys/firmware/efi && test -e "${linux}.efi.signed"; then
+    cat << EOF
+	linux	${rel_dirname}/${basename}.efi.signed root=${linux_root_device_thisversion} ro ${args}
+EOF
+  else
+    cat << EOF
+	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args}
+EOF
+  fi
+  if test -n "${initrd}" ; then
+    # TRANSLATORS: ramdisk isn't identifier. Should be translated.
+    message="$(gettext_printf "Loading initial ramdisk ...")"
+    cat << EOF
+	echo	'$(echo "$message" | grub_quote)'
+EOF
+    initrd_path=
+    for i in ${initrd}; do
+      initrd_path="${initrd_path} ${rel_dirname}/${i}"
+    done
+    cat << EOF
+	initrd	$(echo $initrd_path)
+EOF
+  fi
+  cat << EOF
+}
+EOF
+}
+
+machine=`uname -m`
+case "x$machine" in
+    xx86_64)
+	GENKERNEL_ARCH="x86"
+	list=
+	for i in /boot/vmlinuz-* /vmlinuz-* /boot/kernel-* ; do
+	    if grub_file_is_not_garbage "$i" ; then list="$list $i" ; fi
```

Comment:
Will this generate one top-level menu entry for each kernel found under /boot? How about adding an upper limit of the numbers of entries to generate?

Reply:
Yes it will. This is a more or less a copy of what the 10_linux grub-mkconfig helper script does. They also do not limit the number of entries. Since the usual scenario is installing the ACRN board inspector on a fresh install, this will only lead to only one additional top-level menu entry. It virtually doubles the top-level boot entries you had before installing ACRN board inspector. Since this is the way it is also done in other grub helper scripts and considering the main use case I do not think we need any limitation. BTW, after obtaining the board configuration, the board inspector should be deinstalled again before continuing to prepare the system for ACRN!

Reply:
One difference between 10_linux/25_linux_acrn and this script is that the later does not create any submenu.

But your point also makes sense: if it is a freshly installed distribution, doubling the top-level entries is not a major concern.

May I know why the board inspector shall be removed?

Reply:
Does it make sense to have the board inspector on a ACRN enabled system?